Keyboard accessibility improvements

jenswittmann opened the pull request:
### What does it do?
Make the MODX manager interface more accessible when using the keyboard (WIP 🧑‍🏭 ).

https://github.com/user-attachments/assets/ced65b4c-26a7-44d7-8527-60441d648559

### How to test
Minify CSS and compress JS via grunt first. Use the `Tab` key to navigate through the Manager UI. Use the `Esc` key to close the subnavigation. Use the tab key to navigate through the tabs.

### Related issue(s)/PR(s)
See #16612

modxcommunity commented:
This pull request has been mentioned on **MODX Community**. There might be relevant details there:

https://community.modx.com/t/modx-manager-accessibilty/7990/3

opengeek commented:
@jenswittmann — thanks for this contribution, Jens. When submitting PRs, please do NOT commit changes to the compiled CSS or minimized JS files produced during the grunt build. You can indicate that a grunt build is required to test the changes in the PR section **How to test**. Let me know if you have any questions.

opengeek commented:
Thank you for your contribution! Before your pull request can be reviewed, please [sign the MODX Contributor License Agreement (CLA)](https://modx.com/community/contribute/cla/). This is to ensure MODX has your written permission to distribute projects containing your contributions under the appropriate license.

Please make sure the CLA has been signed for GitHub user(s):  @jenswittmann 

Once you've signed, please reply with a comment letting us know so we can verify.

jenswittmann commented:
> Thank you for your contribution! Before your pull request can be reviewed, please [sign the MODX Contributor License Agreement (CLA)](https://modx.com/community/contribute/cla/). This is to ensure MODX has your written permission to distribute projects containing your contributions under the appropriate license.
> 
> Please make sure the CLA has been signed for GitHub user(s): @jenswittmann
> 
> Once you've signed, please reply with a comment letting us know so we can verify.

@opengeek i signed the CLA 👍🏽

jenswittmann committed:
A11y: add _a11y.scss and reset focus outline in manager #16612

override .ext-webkit *:focus{ outline: none !important; }

jenswittmann committed:
A11y: add keyboard navigation demo for manager main nav #16612

jenswittmann commented:
> @jenswittmann — thanks for this contribution, Jens. When submitting PRs, please do NOT commit changes to the compiled CSS or minimized JS files produced during the grunt build. You can indicate that a grunt build is required to test the changes in the PR section **How to test**. Let me know if you have any questions.

@opengeek thanks for the info. I remove the minified and compressed files and add a hint on **how to test**.

theboxer reviewed (approved):
Awesome addition, thank you. Works as described

jenswittmann committed:
A11y: optimize keyboard navigation for manager main nav #16612

jenswittmann commented:
After testing the PR I found some problems and tried to fix them.

- The whole [navigation is marked as a landmark](https://github.com/modxcms/revolution/pull/16613/files#diff-3ebb225f79f2da772a42ba18b8d7c7b2c567928a9d686cb8770a45164baa1810R82-R116) with `<section aria-label="Navigation">`.
- In general, the subnavigation should be closed when you leave focus on the last item. The subnavigation is [placed in the footer](https://github.com/modxcms/revolution/pull/16613/files#diff-8ac5eb40df4e15acb8ed4bb13cce3beb25ba48452bb5904106d25ac5073ee5b6L10-L11) for some reason (?). So when a subnavigation is open, it is the last DOM element on the page. So focus after the last item leaves the page and brings focus to the browser UI (e.g. Current Tab). There is no way to bring the focus back to the opener button. So [I move the subnavigation](https://github.com/modxcms/revolution/pull/16613/files#diff-3ebb225f79f2da772a42ba18b8d7c7b2c567928a9d686cb8770a45164baa1810R109-R114) from the footer to the header.
- When you leave the focus on the last subnavigation item, the subnavigation is now closed and the focus is on the opener button.
- It is now possible to use the keyboard to navigate the sub-navigation.
- For now, I have set the [focus outline online for the navigation](https://github.com/modxcms/revolution/pull/16613/files#diff-9ec04352d073be82bd7f67f7bcd548835adb490d6df90403de61b88d40fbd680R1-R4). On the other UI elements there are too many side effects at the moment. This should be fixed component by component.

Here is a small demo of the new update:

https://github.com/user-attachments/assets/ced65b4c-26a7-44d7-8527-60441d648559

opengeek commented:
@jenswittmann — we had an issue with the CLA form when you attempted to sign it, so it did not record the action. Could you do that again when you get a chance? Sorry for the inconvenience.

jenswittmann commented:
> @jenswittmann — we had an issue with the CLA form when you attempted to sign it, so it did not record the action. Could you do that again when you get a chance? Sorry for the inconvenience.

I submit the CLA form again. Is it ok now?

opengeek commented:
> > @jenswittmann — we had an issue with the CLA form when you attempted to sign it, so it did not record the action. Could you do that again when you get a chance? Sorry for the inconvenience.
> 
> I submit the CLA form again. Is it ok now?

It is! Thank you!

theboxer commented:
@jenswittmann I'm not sure if this was also an issue before the new update, but menu items that have JS handler (clear cache, delete locks, flush your permissions, etc.) won't trigger when hitting enter.

jenswittmann committed:
A11y: add missing href attribute for manager main nav to ensure onclick fires on keyboard navigation #16612

jenswittmann commented:
@theboxer thanks for this bug! I think its an older bug. Now i add a `href` attribute to the links to properly fire the `onclick` event https://github.com/modxcms/revolution/pull/16613/commits/b99360318dbb9e3a874e1b46958cd70aa566f021 ✅ 

Can you check it again?

theboxer commented:
@jenswittmann thank you! Now those items trigger fine.

I noticed one more thing, once you're under sub menu, you can't escape from it using `esc` key, you have to finish going through it. You think you could adjust this as well?

jenswittmann committed:
A11y: add keydown escape for manager main nav #16612

jenswittmann commented:
@theboxer i restore the escape function to go back on the root button 96e2542 ✅ 

The way the current navigation is structured, it will be difficult to optimise the keyboard navigation even more. So i think this is the best what we can get for now. Have a look at the example "Dropdown Menus" from [Floating UI](https://floating-ui.com/). I thinks this is the goal, but its based on React. A complete rewrite of the Navigation in Vanilla JS with some help of [AlpineJS Focus Plugin](https://alpinejs.dev/plugins/focus#focus-magic) could be a good choice 🤔 #16612

theboxer commented on manager/assets/modext/core/modx.layout.js:
```diff
@@ -465,10 +475,13 @@ Ext.extend(MODx.Layout, Ext.Viewport, {
             });
-            buttons[i].onclick = function(e) {
+            var _this = this;
+            var openSubnav =
```
`var openSubnav =` seems to be invalid :)

jenswittmann replied:
@theboxer i also removed this bug 4be7479a8b12f52d193169d76f2c77e8f9c80bb6 ✅

theboxer commented on manager/assets/modext/core/modx.layout.js:
```diff
@@ -465,10 +475,13 @@ Ext.extend(MODx.Layout, Ext.Viewport, {
                 }
             });
-            buttons[i].onclick = function(e) {
+            var _this = this;
```
`this` is already assigned to `el`, so making new variable `_this` is redundant, you can use `el`. 

Btw. I'm not a huge fan of passing scope this way, it can be completely avoided for example using arrow functions, but it's not the only case in this file where `this` is assigned, so I guess it doesn't matter that much right now.

jenswittmann replied:
@theboxer thanks for the hint. I fix it 1c92247 ✅

jenswittmann committed:
A11y: make modx-tabs focusable #16612

jenswittmann committed:
A11y: optimie JS _this/el #16612

jenswittmann committed:
A11y: bugfix JS var openSubnav = #16612

jenswittmann committed:
A11y: optimize focus for checkboxes #16612

jenswittmann commented:
Thanks to @theboxer and @Ibochkarev for their help 👍🏽 
I have fixed some issues and moved PR #16614 to this for better handling. This PR is now for some global accessibility optimisations, not just navigation.

The checkboxes now also get an outline when focused via the keyboard. ec4e1d9f71bcdf48f3677fa61d51d40c6b36932c ✅

jenswittmann committed:
A11y: use <a> element for modx-button to be focusable in resource tree for create/refresh #16612

jenswittmann commented:
I have changed the markup for the `modx-button` component. The create and refresh buttons are now keyboard focusable and triggerable in the resource tree f903a7315fb762405eb5aed57b7f44d9b65bd90c ✅

theboxer commented:
@jenswittmann pls let me know once this PR will be complete so I can re-review it

jenswittmann commented:
@theboxer I did another keyboard test. Now it should be better to focus on the main UI elements. All other tweaks like windows, tables, landmark/section movement and so on should be in a separate PR for better handling. Maybe this PR can be included in the next 3.0.6 release instead of 3.1.0 🤔 

https://github.com/user-attachments/assets/6f0a1238-9f53-44bf-93dc-083d389df92c

jenswittmann committed:
A11y: add skiplinks and correct tabindex order #16612

jenswittmann commented:
@theboxer I found some problems with the tab order and added some skiplinks to avoid long tab adventures aac69f5 ✅ 

You can now test everything for the next MODX release 👍🏽 

https://github.com/user-attachments/assets/d83545bf-d581-4458-b854-9e828c6fea0f

theboxer reviewed (approved):
Looks good to me. Thx for all the a11y improvements!

opengeek committed:
grunt build

Jako commented:
This patch shows a strange white line in Ext.Msg buttons. See the attached screenshot.

<img width="87" alt="Bildschirmfoto 2025-03-12 um 13 48 41" src="https://github.com/user-attachments/assets/5a0cbe94-0f74-4555-84e2-acaa6c13c4e1" />

jenswittmann commented:
Thank you @Jako! Its because the button is correctly outlined on focus, but the actual visual button is a wrapper:

![image](https://github.com/user-attachments/assets/42c357c9-421d-440e-ba4a-cd92a978a9e8)

So there must be an outline for the wrapper when `:focus-within`. But this needs more testing. I keep an eye on it!

Jako commented:
Sure but the main button in a Ext.Msg window has the focus when opened without any keyboard use. That destroys the design a bit.

jenswittmann commented:
@Jako that is correct. Because when opening a modal, the focus should be on the first focusable element. We can fine-tune this with `:focus-visible`, then the focus is set, but only visually when using the keyboard. I think this solves the design problem?

Jako commented:
The rule that shows this outline when the window is opened is `.ext-webkit *:focus-visible`. And it is directly visible. So `:focus-visible` maybe does not help and the patch has to go into ExtJS.
